# Task to 4.8: Grouping Data & Aggregating Variables
# Contents
## 1. Import libraries&data sets
## 2. Grouping Data with pandas with groupby()
## 3. Aggregating Data with agg()
## 4. Performing Multiple Aggregations
## 5. Aggregating Data with transform()
## 6. Deriving Columns with loc()
## 7. Step 2. Find the aggregated mean of the “customer_order_count" (earlier renamed from "order_number”) column grouped by “department_id” for the entire dataframe.
## 8. Step 3. How do the results for the entire dataframe differ from those of the subset? 
## 9. Step 4. Creating a loyalty flag for existing customers using the transform() and loc() functions.
## 10. Step 5. Find out if there’s a difference between the spending habits of the three types of customers.
## 11. Step 6. Create a spending flag for each user based on the average price across all their orders using the following criteria:
## 12. Step 7.Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. 


## 1. Import libraries&data sets

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Create project folder path
path=r'/Users/yevgeniyaem/Documents/Weiterbildung Data Analytics/11-2024 Instacart Basket Analysis'

In [5]:
path

'/Users/yevgeniyaem/Documents/Weiterbildung Data Analytics/11-2024 Instacart Basket Analysis'

In [7]:
# Import "orders_products_merged.pkl"
ords_prods_merged=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [8]:
# Create a subset with 1mil rows
df = ords_prods_merged[:1000000]

In [9]:
# Check the shape
df.shape

(1000000, 14)

In [10]:
# Double-check the column names
df.head(15)

order_id  user_id  customers_order_count  orders_day_of_week  \
0   2539329        1                      1                   2   
1   2539329        1                      1                   2   
2   2539329        1                      1                   2   
3   2539329        1                      1                   2   
4   2539329        1                      1                   2   
5   2398795        1                      2                   3   
6   2398795        1                      2                   3   
7   2398795        1                      2                   3   
8   2398795        1                      2                   3   
9   2398795        1                      2                   3   
10  2398795        1                      2                   3   
11   473747        1                      3                   3   
12   473747        1                      3                   3   
13   473747        1                      3                   3   
14   473747        1                      3                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   8                     0.0       14084                  2   
2                   8                     0.0       12427                  3   
3                   8                     0.0       26088                  4   
4                   8                     0.0       26405                  5   
5                   7                    15.0         196                  1   
6                   7                    15.0       10258                  2   
7                   7                    15.0       12427                  3   
8                   7                    15.0       13176                  4   
9                   7                    15.0       26088                  5   
10                  7                    15.0       13032                  6   
11                 12                    21.0         196                  1   
12                 12                    21.0       12427                  2   
13                 12                    21.0       10258                  3   
14                 12                    21.0       25133                  4   

    reordered                             product_name  aisle_id  \
0           0                                     Soda        77   
1           0  Organic Unsweetened Vanilla Almond Milk        91   
2           0                      Original Beef Jerky        23   
3           0               Aged White Cheddar Popcorn        23   
4           0         XL Pick-A-Size Paper Towel Rolls        54   
5           1                                     Soda        77   
6           0                               Pistachios       117   
7           1                      Original Beef Jerky        23   
8           0                   Bag of Organic Bananas        24   
9           1               Aged White Cheddar Popcorn        23   
10          0                    Cinnamon Toast Crunch       121   
11          1                                     Soda        77   
12          1                      Original Beef Jerky        23   
13          1                               Pistachios       117   
14          0                    Organic String Cheese        21   

    department_id  prices _merge  
0               7     9.0   both  
1              16    12.5   both  
2              19     4.4   both  
3              19     4.7   both  
4              17     1.0   both  
5               7     9.0   both  
6              19     3.0   both  
7              19     4.4   both  
8               4    10.3   both  
9              19     4.7   both  
10             14     4.0   both  
11              7     9.0   both  
12             19     4.4   both  
13             19     3.0   both  
14             16     8.6   both

## 2. Grouping Data with pandas with groupby()
## 1. Split the data into groups based on the given criteria. 
## 2. Apply a function to each group separately. 
## 3. Combine the results into a dataframe or alternative data structure or create a new column in the current dataframe.

In [11]:
# Group df by "product_name"
df.groupby('product_name')

### The function created a pandas object
### However, the output is not visible yet
### Something else needs to be done, like aggregating the data or applying a function, before seeing the results
### groupby() should always be used as part of a series of steps

## 3. Aggregating Data with agg()
### Aggregating functions create summaries of selected columns and store these summary values in new columns
### These summaries usually take the form of a descriptive statistic

In [17]:
# 1. Split the data into groups based on “department_id.”
# 2. Apply the agg() function to each group to obtain the mean values for the “customers_order_count" (renamed from order_number) column.
df.groupby('department_id').agg({'customers_order_count': ['mean']})

customers_order_count
                               mean
department_id                      
1                         14.800024
2                         17.091743
3                         17.913544
4                         17.893092
5                         15.214270
6                         15.382135
7                         17.694027
8                         16.458105
9                         15.957363
10                        20.091818
11                        16.482026
12                        15.615061
13                        16.484023
14                        17.524632
15                        15.691875
16                        18.014071
17                        16.150593
18                        19.602850
19                        17.631340
20                        17.138607
21                        21.956893

In [19]:
# There are some aggregations that can be conducted without use of the agg() function. 
# For instance, with a command that uses the mean() function to achieve the same results
df.groupby('department_id')['customers_order_count'].mean()

department_id
1     14.800024
2     17.091743
3     17.913544
4     17.893092
5     15.214270
6     15.382135
7     17.694027
8     16.458105
9     15.957363
10    20.091818
11    16.482026
12    15.615061
13    16.484023
14    17.524632
15    15.691875
16    18.014071
17    16.150593
18    19.602850
19    17.631340
20    17.138607
21    21.956893
Name: customers_order_count, dtype: float64

## 4. Performing Multiple Aggregations

In [21]:
# Produce multiply statistics at the same time: mean, max and min
df.groupby('department_id').agg({'customers_order_count': ['mean', 'min', 'max']})

customers_order_count        
                               mean min max
department_id                              
1                         14.800024   1  99
2                         17.091743   1  98
3                         17.913544   1  99
4                         17.893092   1  99
5                         15.214270   1  99
6                         15.382135   1  99
7                         17.694027   1  99
8                         16.458105   1  91
9                         15.957363   1  99
10                        20.091818   1  99
11                        16.482026   1  99
12                        15.615061   1  99
13                        16.484023   1  99
14                        17.524632   1  99
15                        15.691875   1  99
16                        18.014071   1  99
17                        16.150593   1  99
18                        19.602850   1  99
19                        17.631340   1  99
20                        17.138607   1  99
21                        21.956893   1  97

## 5.  Aggregating Data with transform()
### We'll be creating 'loyalty' flag if our dataframe
### 'Loyalty' customers are those who come back time and time again to use the service or buy products
### We'll be locating these customers so that the business strategy team can employ some kind of bonus point program for them
### We'll using the 'transform()' function, which will create a new column containing the maximum frequency of the 'customers_order_count' (renamed from order_number) column
### Then, using the loc() function, a second column will be created containing a flag designating whether a customer is 'loyal' or not

In [ ]:
# Creating a "loyalty flag" column in ords_prods_merged dataframe:
# If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
# If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
# If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [23]:
# 1. Split the data into groups based on the “user_id” column.
# 2. Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
# 3. Create a new column, “max_order,” into which you’ll place the results of your aggregation.
ords_prods_merged['max_order'] = ords_prods_merged.groupby(['user_id'])['customers_order_count'].transform(np.max)

/var/folders/gn/b13h25pn5k1500kc2862p30r0000gp/T/ipykernel_30678/4085896468.py:4: FutureWarning: The provided callable <function max at 0x10ed2f6a0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merged['max_order'] = ords_prods_merged.groupby(['user_id'])['customers_order_count'].transform(np.max)


In [25]:
ords_prods_merged.head(15)

order_id  user_id  customers_order_count  orders_day_of_week  \
0   2539329        1                      1                   2   
1   2539329        1                      1                   2   
2   2539329        1                      1                   2   
3   2539329        1                      1                   2   
4   2539329        1                      1                   2   
5   2398795        1                      2                   3   
6   2398795        1                      2                   3   
7   2398795        1                      2                   3   
8   2398795        1                      2                   3   
9   2398795        1                      2                   3   
10  2398795        1                      2                   3   
11   473747        1                      3                   3   
12   473747        1                      3                   3   
13   473747        1                      3                   3   
14   473747        1                      3                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   8                     0.0       14084                  2   
2                   8                     0.0       12427                  3   
3                   8                     0.0       26088                  4   
4                   8                     0.0       26405                  5   
5                   7                    15.0         196                  1   
6                   7                    15.0       10258                  2   
7                   7                    15.0       12427                  3   
8                   7                    15.0       13176                  4   
9                   7                    15.0       26088                  5   
10                  7                    15.0       13032                  6   
11                 12                    21.0         196                  1   
12                 12                    21.0       12427                  2   
13                 12                    21.0       10258                  3   
14                 12                    21.0       25133                  4   

    reordered                             product_name  aisle_id  \
0           0                                     Soda        77   
1           0  Organic Unsweetened Vanilla Almond Milk        91   
2           0                      Original Beef Jerky        23   
3           0               Aged White Cheddar Popcorn        23   
4           0         XL Pick-A-Size Paper Towel Rolls        54   
5           1                                     Soda        77   
6           0                               Pistachios       117   
7           1                      Original Beef Jerky        23   
8           0                   Bag of Organic Bananas        24   
9           1               Aged White Cheddar Popcorn        23   
10          0                    Cinnamon Toast Crunch       121   
11          1                                     Soda        77   
12          1                      Original Beef Jerky        23   
13          1                               Pistachios       117   
14          0                    Organic String Cheese        21   

    department_id  prices _merge  max_order  
0               7     9.0   both         10  
1              16    12.5   both         10  
2              19     4.4   both         10  
3              19     4.7   both         10  
4              17     1.0   both         10  
5               7     9.0   both         10  
6              19     3.0   both         10  
7              19     4.4   both         10  
8               4    10.3   both         10  
9              19     4.7   both         10  
10             14     4.0   both         10  
11      

In [27]:
# After warning i adjusted the code: 
ords_prods_merged['max_order'] = ords_prods_merged.groupby('user_id')['customers_order_count'].transform('max')

In [29]:
# Check the output
ords_prods_merged.head(25)

order_id  user_id  customers_order_count  orders_day_of_week  \
0   2539329        1                      1                   2   
1   2539329        1                      1                   2   
2   2539329        1                      1                   2   
3   2539329        1                      1                   2   
4   2539329        1                      1                   2   
5   2398795        1                      2                   3   
6   2398795        1                      2                   3   
7   2398795        1                      2                   3   
8   2398795        1                      2                   3   
9   2398795        1                      2                   3   
10  2398795        1                      2                   3   
11   473747        1                      3                   3   
12   473747        1                      3                   3   
13   473747        1                      3                   3   
14   473747        1                      3                   3   
15   473747        1                      3                   3   
16  2254736        1                      4                   4   
17  2254736        1                      4                   4   
18  2254736        1                      4                   4   
19  2254736        1                      4                   4   
20  2254736        1                      4                   4   
21   431534        1                      5                   4   
22   431534        1                      5                   4   
23   431534        1                      5                   4   
24   431534        1                      5                   4   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   8                     0.0       14084                  2   
2                   8                     0.0       12427                  3   
3                   8                     0.0       26088                  4   
4                   8                     0.0       26405                  5   
5                   7                    15.0         196                  1   
6                   7                    15.0       10258                  2   
7                   7                    15.0       12427                  3   
8                   7                    15.0       13176                  4   
9                   7                    15.0       26088                  5   
10                  7                    15.0       13032                  6   
11                 12                    21.0         196                  1   
12                 12                    21.0       12427                  2   
13                 12                    21.0       10258                  3   
14                 12                    21.0       25133                  4   
15                 12                    21.0       30450                  5   
16                  7                    29.0         196                  1   
17                  7                    29.0       12427                  2   
18                  7                    29.0       10258                  3   
19                  7                    29.0       25133                  4   
20                  7                    29.0       26405                  5   
21                 15                    28.0         196                  1   
22                 15                    28.0       12427                  2   
23                 15                    28.0       10258                  3   
24                 15                    28.0       25133                  4   

    reordered                             product_name  aisle_id  \
0           0                                     Soda        77   
1           0  Organic Unsweetened Vanilla 

In [31]:
# Double-check of result
ords_prods_merged.head(100)

order_id  user_id  customers_order_count  orders_day_of_week  \
0   2539329        1                      1                   2   
1   2539329        1                      1                   2   
2   2539329        1                      1                   2   
3   2539329        1                      1                   2   
4   2539329        1                      1                   2   
..      ...      ...                    ...                 ...   
95   738281        2                      4                   2   
96  1673511        2                      5                   3   
97  1673511        2                      5                   3   
98  1673511        2                      5                   3   
99  1673511        2                      5                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   8                     0.0       14084                  2   
2                   8                     0.0       12427                  3   
3                   8                     0.0       26088                  4   
4                   8                     0.0       26405                  5   
..                ...                     ...         ...                ...   
95                 10                     8.0       21150                 13   
96                 11                     8.0       47144                  1   
97                 11                     8.0        5322                  2   
98                 11                     8.0       17224                  3   
99                 11                     8.0       38656                  4   

    reordered                                       product_name  aisle_id  \
0           0                                               Soda        77   
1           0            Organic Unsweetened Vanilla Almond Milk        91   
2           0                                Original Beef Jerky        23   
3           0                         Aged White Cheddar Popcorn        23   
4           0                   XL Pick-A-Size Paper Towel Rolls        54   
..        ...                                                ...       ...   
95          0                            Fire Grilled Steak Bowl        38   
96          0     Unsweetened Original Almond Breeze Almond Milk        91   
97          0  Gluten Free Dark Chocolate Chunk Chewy with a ...         3   
98          0                   Oats & Honey Gluten Free Granola         3   
99          0  Gluten Free Peanut Butter Dark Chocolate Chewy...         3   

    department_id  prices _merge  max_order  
0               7     9.0   both         10  
1              16    12.5   both         10  
2              19     4.4   both         10  
3              19     4.7   both         10  
4              17     1.0   both         10  
..            ...     ...    ...        ...  
95              1     5.9   both         14  
96             16    14.0   both         14  
97             19     2.9   both         14  
98             19     1.6   both         14  
99             19     3.4   both         14  

[100 rows x 15 columns]

In [33]:
# Another way to check the result:
pd.options.display.max_rows = None

In [35]:
ords_prods_merged.head(100)

order_id  user_id  customers_order_count  orders_day_of_week  \
0   2539329        1                      1                   2   
1   2539329        1                      1                   2   
2   2539329        1                      1                   2   
3   2539329        1                      1                   2   
4   2539329        1                      1                   2   
5   2398795        1                      2                   3   
6   2398795        1                      2                   3   
7   2398795        1                      2                   3   
8   2398795        1                      2                   3   
9   2398795        1                      2                   3   
10  2398795        1                      2                   3   
11   473747        1                      3                   3   
12   473747        1                      3                   3   
13   473747        1                      3                   3   
14   473747        1                      3                   3   
15   473747        1                      3                   3   
16  2254736        1                      4                   4   
17  2254736        1                      4                   4   
18  2254736        1                      4                   4   
19  2254736        1                      4                   4   
20  2254736        1                      4                   4   
21   431534        1                      5                   4   
22   431534        1                      5                   4   
23   431534        1                      5                   4   
24   431534        1                      5                   4   
25   431534        1                      5                   4   
26   431534        1                      5                   4   
27   431534        1                      5                   4   
28   431534        1                      5                   4   
29  3367565        1                      6                   2   
30  3367565        1                      6                   2   
31  3367565        1                      6                   2   
32  3367565        1                      6                   2   
33   550135        1                      7                   1   
34   550135        1                      7                   1   
35   550135        1                      7                   1   
36   550135        1                      7                   1   
37   550135        1                      7                   1   
38  3108588        1                      8                   1   
39  3108588        1                      8                   1   
40  3108588        1                      8                   1   
41  3108588        1                      8                   1   
42  3108588        1                      8                   1   
43  3108588        1                      8                   1   
44  2295261        1                      9                   1   
45  2295261        1                      9                   1   
46  2295261        1                      9                   1   
47  2295261        1                      9                   1   
48  2295261        1                      9                   1   
49  2295261        1                      9                   1   
50  2550362        1                     10                   4   
51  2550362        1                     10                   4   
52  2550362        1                     10                   4   
53  2550362        1                     10                   4   
54  2550362        1                     10                   4   
55  2550362        1                     10                   4   
56  2550362        1                     10                   4   
57  2550362        1                     10                   4   
58  2550362        1                     10       

## 6. Deriving Columns with loc()

In [37]:
# Now that the new column is created, we must flag customers who meet the criteria as 'loyal'
ords_prods_merged.loc[ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [38]:
ords_prods_merged.loc[(ords_prods_merged['max_order'] <= 40) & (ords_prods_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [41]:
ords_prods_merged.loc[ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [43]:
# Check value counts for new column
ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [ ]:
# Most customers belongs to Regular customer category

In [45]:
# Check that everything was interpreted correctly and that the right flags were assigned 
ords_prods_merged[['user_id', 'loyalty_flag', 'customers_order_count']].head(60)

user_id      loyalty_flag  customers_order_count
0         1      New customer                      1
1         1      New customer                      1
2         1      New customer                      1
3         1      New customer                      1
4         1      New customer                      1
5         1      New customer                      2
6         1      New customer                      2
7         1      New customer                      2
8         1      New customer                      2
9         1      New customer                      2
10        1      New customer                      2
11        1      New customer                      3
12        1      New customer                      3
13        1      New customer                      3
14        1      New customer                      3
15        1      New customer                      3
16        1      New customer                      4
17        1      New customer                      4
18        1      New customer                      4
19        1      New customer                      4
20        1      New customer                      4
21        1      New customer                      5
22        1      New customer                      5
23        1      New customer                      5
24        1      New customer                      5
25        1      New customer                      5
26        1      New customer                      5
27        1      New customer                      5
28        1      New customer                      5
29        1      New customer                      6
30        1      New customer                      6
31        1      New customer                      6
32        1      New customer                      6
33        1      New customer                      7
34        1      New customer                      7
35        1      New customer                      7
36        1      New customer                      7
37        1      New customer                      7
38        1      New customer                      8
39        1      New customer                      8
40        1      New customer                      8
41        1      New customer                      8
42        1      New customer                      8
43        1      New customer                      8
44        1      New customer                      9
45        1      New customer                      9
46        1      New customer                      9
47        1      New customer                      9
48        1      New customer                      9
49        1      New customer                      9
50        1      New customer                     10
51        1      New customer                     10
52        1      New customer                     10
53        1      New customer                     10
54        1      New customer                     10
55        1      New customer                     10
56        1      New customer                     10
57        1      New customer                     10
58        1      New customer                     10
59        2  Regular customer                      1

## 7. Step 2. Find the aggregated mean of the “customer_order_count" (earlier renamed from "order_number”) column grouped by “department_id” for the entire dataframe.

In [47]:
# Grouping by 'department_id' and 'order_number', and Calculating Mean using agg() Function
ords_prods_merged.groupby('department_id').agg({'customers_order_count': ['mean']})

customers_order_count
                               mean
department_id                      
1                         15.457838
2                         17.277920
3                         17.170395
4                         17.811403
5                         15.215751
6                         16.439806
7                         17.225802
8                         15.340650
9                         15.895474
10                        20.197148
11                        16.170638
12                        15.887671
13                        16.583536
14                        16.773669
15                        16.165037
16                        17.665606
17                        15.694469
18                        19.310397
19                        17.177343
20                        16.473447
21                        22.902379

## 8. Step 3. How do the results for the entire dataframe differ from those of the subset? 
## Overall Trends: The entire dataframe generally has higher mean order numbers across departments, 
## indicating it captures a broader range of customer behaviours than the subset.

## 9. Step 4. Creating a loyalty flag for existing customers using the transform() and loc() functions.

In [85]:
# Aggregating Data with transform() function
ords_prods_merged['max_order'] = ords_prods_merged.groupby(['user_id'])['customers_order_count'].transform('max')

In [57]:
# Checking the Output
ords_prods_merged.head(15)

order_id  user_id  customers_order_count  orders_day_of_week  \
0   2539329        1                      1                   2   
1   2539329        1                      1                   2   
2   2539329        1                      1                   2   
3   2539329        1                      1                   2   
4   2539329        1                      1                   2   
5   2398795        1                      2                   3   
6   2398795        1                      2                   3   
7   2398795        1                      2                   3   
8   2398795        1                      2                   3   
9   2398795        1                      2                   3   
10  2398795        1                      2                   3   
11   473747        1                      3                   3   
12   473747        1                      3                   3   
13   473747        1                      3                   3   
14   473747        1                      3                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   8                     0.0       14084                  2   
2                   8                     0.0       12427                  3   
3                   8                     0.0       26088                  4   
4                   8                     0.0       26405                  5   
5                   7                    15.0         196                  1   
6                   7                    15.0       10258                  2   
7                   7                    15.0       12427                  3   
8                   7                    15.0       13176                  4   
9                   7                    15.0       26088                  5   
10                  7                    15.0       13032                  6   
11                 12                    21.0         196                  1   
12                 12                    21.0       12427                  2   
13                 12                    21.0       10258                  3   
14                 12                    21.0       25133                  4   

    reordered                             product_name  aisle_id  \
0           0                                     Soda        77   
1           0  Organic Unsweetened Vanilla Almond Milk        91   
2           0                      Original Beef Jerky        23   
3           0               Aged White Cheddar Popcorn        23   
4           0         XL Pick-A-Size Paper Towel Rolls        54   
5           1                                     Soda        77   
6           0                               Pistachios       117   
7           1                      Original Beef Jerky        23   
8           0                   Bag of Organic Bananas        24   
9           1               Aged White Cheddar Popcorn        23   
10          0                    Cinnamon Toast Crunch       121   
11          1                                     Soda        77   
12          1                      Original Beef Jerky        23   
13          1                               Pistachios       117   
14          0                    Organic String Cheese        21   

    department_id  prices _merge  max_order  loyalty_flag  
0               7     9.0   both         10  New customer  
1              16    12.5   both         10  New customer  
2              19     4.4   both         10  New customer  
3              19     4.7   both         10  New customer  
4              17     1.0   both         10  New customer  
5               7     9.0   both         10  New customer  
6              19     3.0   both         10  New customer  
7              19     4.4   both         10  New customer  
8               4   

In [59]:
# Creating a 'Loyalty' Label Flag using loc function
ords_prods_merged.loc[ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [61]:
ords_prods_merged.loc[(ords_prods_merged['max_order'] <= 40) & (ords_prods_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [63]:
ords_prods_merged.loc[ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [65]:
# Check the value counts
ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [67]:
# Check the Updated output
ords_prods_merged.head(15)

order_id  user_id  customers_order_count  orders_day_of_week  \
0   2539329        1                      1                   2   
1   2539329        1                      1                   2   
2   2539329        1                      1                   2   
3   2539329        1                      1                   2   
4   2539329        1                      1                   2   
5   2398795        1                      2                   3   
6   2398795        1                      2                   3   
7   2398795        1                      2                   3   
8   2398795        1                      2                   3   
9   2398795        1                      2                   3   
10  2398795        1                      2                   3   
11   473747        1                      3                   3   
12   473747        1                      3                   3   
13   473747        1                      3                   3   
14   473747        1                      3                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   8                     0.0       14084                  2   
2                   8                     0.0       12427                  3   
3                   8                     0.0       26088                  4   
4                   8                     0.0       26405                  5   
5                   7                    15.0         196                  1   
6                   7                    15.0       10258                  2   
7                   7                    15.0       12427                  3   
8                   7                    15.0       13176                  4   
9                   7                    15.0       26088                  5   
10                  7                    15.0       13032                  6   
11                 12                    21.0         196                  1   
12                 12                    21.0       12427                  2   
13                 12                    21.0       10258                  3   
14                 12                    21.0       25133                  4   

    reordered                             product_name  aisle_id  \
0           0                                     Soda        77   
1           0  Organic Unsweetened Vanilla Almond Milk        91   
2           0                      Original Beef Jerky        23   
3           0               Aged White Cheddar Popcorn        23   
4           0         XL Pick-A-Size Paper Towel Rolls        54   
5           1                                     Soda        77   
6           0                               Pistachios       117   
7           1                      Original Beef Jerky        23   
8           0                   Bag of Organic Bananas        24   
9           1               Aged White Cheddar Popcorn        23   
10          0                    Cinnamon Toast Crunch       121   
11          1                                     Soda        77   
12          1                      Original Beef Jerky        23   
13          1                               Pistachios       117   
14          0                    Organic String Cheese        21   

    department_id  prices _merge  max_order  loyalty_flag  
0               7     9.0   both         10  New customer  
1              16    12.5   both         10  New customer  
2              19     4.4   both         10  New customer  
3              19     4.7   both         10  New customer  
4              17     1.0   both         10  New customer  
5               7     9.0   both         10  New customer  
6              19     3.0   both         10  New customer  
7              19     4.4   both         10  New customer  
8               4   

## 10. Step 5. Find out if there’s a difference between the spending habits of the three types of customers.
## Using loyalty flag and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). 
## Determine whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers

In [69]:
# Determining basic statistics of spending habits by Loyalty Category
ords_prods_merged.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

In [70]:
loyalty_stats = ords_prods_merged.groupby('loyalty_flag')['prices'].describe() 
print(loyalty_stats)

                       count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10284093.0  10.386336  328.017787  1.0  4.2  7.4  11.2   
New customer       6243990.0  13.294670  597.560299  1.0  4.2  7.4  11.3   
Regular customer  15876776.0  12.495717  539.720919  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    99999.0  
New customer      99999.0  
Regular customer  99999.0  


In [72]:
# Determinign the Median of prices grouped by loyalty_flag
ords_prods_merged.groupby('loyalty_flag')['prices'].agg(['count', 'mean', 'median'])

count       mean  median
loyalty_flag                                 
Loyal customer    10284093  10.386336     7.4
New customer       6243990  13.294670     7.4
Regular customer  15876776  12.495717     7.4

#### Loyal customers tend to purchase lower-cost items, with an average price of $10.39.
#### Maybe because their shop more often so they know the product range better and can use cheaper prices. Also maybe they take part in loyalty programs or offers as loyal customers.

#### New customers spend a little bit more, with a mean cost of $13.29. 
#### Because unlike to loyal customers they are not so familiar with the products sortiment and didn't participate in special offers or loyalty programms.

#### Regular customers fall in the middle with averaging $12.50 per item. 

#### All groups share a minimum price of 1$, but the extremely high maximum cost of $99999 suggests possible outliers or data errors.

### Insights: 
#### New customers display the highest mean price, suggesting marketing efforts could target them with premium products to retain their interest.
#### Loyal customers have a slightly lower average price, potentially focusing more on cost-effective options or regular purchases.
#### Across all groups, the maximum price indicates a standard upper limit for product pricing, hinting at opportunities to promote high-value items more effectively.
#### This analysis can guide targeted marketing strategies based on spending patterns across customer categories.

## 11. Step 6. Create a spending flag for each user based on the average price across all their orders using the following criteria:
## If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
## If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [139]:
# 1a. Split the data into groups based on the “user_id” column.
# 1b. Apply the transform() function on the “prices” column to generate the mean price for each user.
# 1c. Create a new column, “avg_user_price,” into which you’ll place the results of your aggregation.
ords_prods_merged['avg_user_price'] = ords_prods_merged.groupby(['user_id'])['prices'].transform('mean')

In [113]:
# Checking Output
ords_prods_merged[['user_id', 'order_id', 'avg_user_price']].head(20)

user_id order_id  avg_user_price
0         1  2539329        6.367797
1         1  2539329        6.367797
2         1  2539329        6.367797
3         1  2539329        6.367797
4         1  2539329        6.367797
5         1  2398795        6.367797
6         1  2398795        6.367797
7         1  2398795        6.367797
8         1  2398795        6.367797
9         1  2398795        6.367797
10        1  2398795        6.367797
11        1   473747        6.367797
12        1   473747        6.367797
13        1   473747        6.367797
14        1   473747        6.367797
15        1   473747        6.367797
16        1  2254736        6.367797
17        1  2254736        6.367797
18        1  2254736        6.367797
19        1  2254736        6.367797

In [115]:
# 2. Assign the Spending Flag for High Spenders&Low Spender
ords_prods_merged.loc[ords_prods_merged['avg_user_price'] >= 10, 'spending_flag'] = 'High spender'
ords_prods_merged.loc[ords_prods_merged['avg_user_price'] < 10, 'spending_flag'] = 'Low spender'

In [117]:
# Examining Value Count on 'spending_flag' Column
ords_prods_merged['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [121]:
# Checking Output
ords_prods_merged[['user_id', 'order_id', 'avg_user_price']].tail(20)

user_id order_id  avg_user_price
32404839   206209   626363        7.058915
32404840   206209   626363        7.058915
32404841   206209   626363        7.058915
32404842   206209   626363        7.058915
32404843   206209   626363        7.058915
32404844   206209   626363        7.058915
32404845   206209   626363        7.058915
32404846   206209   626363        7.058915
32404847   206209   626363        7.058915
32404848   206209   626363        7.058915
32404849   206209   626363        7.058915
32404850   206209  2977660        7.058915
32404851   206209  2977660        7.058915
32404852   206209  2977660        7.058915
32404853   206209  2977660        7.058915
32404854   206209  2977660        7.058915
32404855   206209  2977660        7.058915
32404856   206209  2977660        7.058915
32404857   206209  2977660        7.058915
32404858   206209  2977660        7.058915

In [131]:
# Delete avg_order_price and _merge columns
ords_prods_merged = ords_prods_merged.drop(columns=['avg_order_price', '_merge'])

In [133]:
# Check the result
ords_prods_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 17 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 int64  
 2   customers_order_count   int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
 6   product_id              int64  
 7   add_to_cart_order       int64  
 8   reordered               int64  
 9   product_name            object 
 10  aisle_id                int64  
 11  department_id           int64  
 12  prices                  float64
 13  max_order               int64  
 14  loyalty_flag            object 
 15  spending_flag           object 
 16  avg_user_price          float64
dtypes: float64(3), int64(10), object(4)
memory usage: 4.1+ GB


In [135]:
### We see that "Low spender" represent with 31770614 customers the biggest group, the "High spender" only includes 634245 customers.

## 12. Step 7.Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. 
# The criteria for the flag should be as follows:
## If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
## If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
## If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [141]:
# 1a. Split the data into groups based on the “user_id” column.
# 1b. Apply the transform() function on the “days_since_prior_order” column to generate the median for each user.
# 1c. Create a new column, “median_days_between_orders” into which you’ll place the results of your aggregation.
ords_prods_merged['median_days_between_orders'] = ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [145]:
# Checking outout
ords_prods_merged[['user_id', 'median_days_between_orders']].head()

user_id  median_days_between_orders
0        1                        20.0
1        1                        20.0
2        1                        20.0
3        1                        20.0
4        1                        20.0

In [157]:
# 2.  Assign the Frequency Flag for Non-frequent customer, regular customer and frequent customer.
ords_prods_merged.loc[ords_prods_merged['median_days_between_orders'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [165]:
ords_prods_merged.loc[(ords_prods_merged['median_days_between_orders'] > 10) & 
                      (ords_prods_merged['median_days_between_orders'] <= 20), 
                      'frequency_flag'] = 'Regular customer'

In [161]:
ords_prods_merged.loc[ords_prods_merged['median_days_between_orders'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [167]:
# Examining Value Count on 'frequency_flag' Column
ords_prods_merged['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: count, dtype: int64

In [169]:
### We have 2279659 frequent customers, 6821472 regulat customers and 2686728 non-frequent customers.

In [175]:
# Checking Relevant Columns of Updated Dataframe
ords_prods_merged[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'frequency_flag']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0         1                     0.0                        20.0   
1         1                     0.0                        20.0   
2         1                     0.0                        20.0   
3         1                     0.0                        20.0   
4         1                     0.0                        20.0   
5         1                    15.0                        20.0   
6         1                    15.0                        20.0   
7         1                    15.0                        20.0   
8         1                    15.0                        20.0   
9         1                    15.0                        20.0   
10        1                    15.0                        20.0   
11        1                    21.0                        20.0   
12        1                    21.0                        20.0   
13        1                    21.0                        20.0   
14        1                    21.0                        20.0   

      frequency_flag  
0   Regular customer  
1   Regular customer  
2   Regular customer  
3   Regular customer  
4   Regular customer  
5   Regular customer  
6   Regular customer  
7   Regular customer  
8   Regular customer  
9   Regular customer  
10  Regular customer  
11  Regular customer  
12  Regular customer  
13  Regular customer  
14  Regular customer

In [ ]:
# Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.
ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_grouped.pkl'))